---
title: "10x Visium data exploration"
author: "Helena L. Crowell"
date: "`r format(Sys.Date(), '%b %m, %Y')`"
output: 
    BiocStyle::html_document:
        toc: true
vignette: >
    %\VignetteIndexEntry{Visium Spatial Data Workflow}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
    chunk_output_type: console
---

<style type="text/css"> .smaller { font-size: 10px } </style>

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = TRUE, autodep = TRUE, cache.lazy = FALSE)
```


# Installation

The SpatialExperiment is available via Bioconductor.

```{r, eval=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("SpatialExperiment")
```

Load the package as follows:

```{r message = FALSE, warning = FALSE}
library(SpatialExperiment)
```

# Summary

A brief summary of the package's current structure is outlined below. To give some examples, we load an example `SpatialExperiment` containing 10X Visium spatial gene expression data of two serial mouse brain sections (Sagittal-Posterior) available [here](https://support.10xgenomics.com/spatial-gene-expression/datasets).

```{r}
dir <- system.file(
    file.path("extdata", "10xVisium"), 
    package = "SpatialExperiment")
   
sample_ids <- c("section1", "section2")
samples <- file.path(dir, sample_ids)

(ve <- read10xVisium(samples, sample_ids, type="sparse", 
    data="raw", images = c("lowres"), load = FALSE))
```

## The `SpatialExperiment` class {- .tabset}

### Spatial data {-}

<!-- Spatial data are stored in the `colData` and must contain: -->

<!-- * a character `sample_id` column giving unique sample identifiers -->
<!-- * a logical `in_tissue` column indicating whether or not an observation was mapped to tissue -->
<!-- * a two- or three-column numeric `xy_coords` matrix storing spatial coordinates -->

TBW 

```{r}
head(spatialData(ve))
```
  
### Image data {-} 
  
Image-related data are stored in the `int_metadata()$imgData` field as a `DataFrame` with the following columns:

* `sample_id` and `image_id` specifying the image's sample and image identifier
* `data`: a list of `SpatialImage`s containing the image's `raster`, path or URL
* `width` and `height` giving the image's dimension in pixel
* `scaleFactor` used to re-scale spatial coordinates according to the image's resolution 

```{r}
(df <- imgData(ve))
```

### Getters, setters & methods {-}

* `scaleFactors()` retrieves scale factors
* `imgData(), imgData() <- value` to get/set the `imgData` `DataFrame`
* `getImg()` retrieves a single or list of `SpatialImage`(s) from the `imgData`
* `add/rmvImg()` can be used to create/eliminate images in the `imgData` 
* the `read10xVisium()` constructor can be used to create a `SpatialExperiment` from 10x Visium spatial gene expression data in one line

## The `SpatialImage` class

The `SpatialImage` class hierarchy provides representations of images from a variety of sources, and is used by the `SpatialExperiment` class to manage the loading of images across multiple studies. The class is divided into the `Loaded/Stored/RemoteSpatialImage` class, allowing to realized images into memory fully, or on request from a locally saved file or remotely hosted URL.

```{r}
getImg(ve)
```

* `imgRaster` accesses the `raster` object associated with the input SPI
* `imgSource` accesses the path/URL associated with a `Stored/RemoteSpatialImage`

# The `SpatialExperiment` class

The `SpatialExperiment` class extends the `SingleCellExperiment` class by requiring specific fields to be present in the object's `colData` and `int_metadata`. These aim to accommodate spatially and image related data", "respectively. 

# Reading 10X Visium data

The 10X Genomics' *CellRanger* pipeline will process the data using standard output file formats that are saved, for each sample, in a single directory `/<sample>/outs/` of the following structure: 

```{bash eval = FALSE}
sample 
|—outs 
··|—raw/filtered_feature_bc_matrix.h5 
··|—raw/filtered_feature_bc_matrix 
····|—barcodes.tsv 
····|—features.tsv 
····|—matrix.mtx 
··|—spatial 
····|—scalefactors_json.json 
····|—tissue_lowres_image.png 
····|—tissue_positions_list.csv 
```

We can load these data into a `SpatialExperiment` using the `read10xVisium` function, which will read in all relevant information, including the count data, spatial coordinates, scale factors, and images:

```{r}
dir <- system.file(
    file.path("extdata", "10xVisium"), 
    package = "SpatialExperiment")

sample_ids <- c("section1", "section2")
samples <- file.path(dir, sample_ids)

list.files(samples[1])
list.files(file.path(samples[1], "spatial"))

(ve <- read10xVisium(samples, sample_ids,
    type="sparse",
    data="raw",
    images = "lowres", # specify which image(s) to include
    load = TRUE))      # specify whether or not to load image(s)
```

# Spatial data

Sample identifiers, indication of whether or not an observation was mapped to the tissue, as well as spatial coordinates are stored inside the `colData`.

```{r}
# tabulate number of spots mapped to tissue
table(
    in_tissue = inTissue(ve),
    sample_id = ve$sample_id)
```

```{r}
# view spatial coordinates
head(spatialData(ve))
```


```{r}
head(spatialCoords(ve))
```

# Image data

All image related data are stored inside the `int_metadata`'s `imgData` field as `DataFrame` of the following structure: 

* each row corresponds to one image for a given sample and with a given unique image identifier (e.g. its resolutions)
* for each image, columns specify:
  * which `sample_id` the image belongs to
  * a unique `image_id` in order to accommodate multiple images for a given sample (e.g. of different resolutions)
  * the image's `data` (a `SpatialImage` object)
  * the `scaleFactor` that adjusts pixel positions of the original, full-resolution image to pixel positions in the image

The `imgData()` accessor can be used to retrieve the image data stored within the object:

```{r}
imgData(ve)
```

## The `SpatialImage` class

Images are stored inside the `data` field of the `imgData` as a list of `SpatialImage`s. Each image may be of one of the following sub-classes:

* `LoadedSpatialImage`
  * represents an image that is fully realized into memory as a `raster` object
  * `@image` contains a `raster` object: a matrix of RGB colors for each pixel in the image
* `StoredSpatialImage`: 
  * represents an image that is stored in a local file (e.g., as a .png, .jpg or .tif), and loaded into memory only on request
  * `@path` specifies a local file from which to retrieve the image
* `RemoteSpatialImage`
  * represents an image that is remotely hosted (under some URL), and retrieved only on request
  * `@url` specifies where to retrieve the image from

A `SpatialImage` can be accessed using `getImg()`, or retrieved directly from the `imgData()`:

```{r}
(spi <- getImg(ve))
identical(spi, imgData(ve)$data[[1]])
```

Data available in an object of class `SpatialImage` may be accessed via the `imgRaster()` and `imgSource()` accessors:

```{r fig.small = TRUE}
plot(imgRaster(spi))
```

## Adding/removing images

Images entries may be added or removed from a `SpatialExperiment`'s `imgData` `DataFrame` using `addImg()` and `rmvImg()`, respectively.

Besides a path or URL to source the image from and a numeric scale factor, `addImg()` requires specification of the `sample_id` the new image belongs to, and an `image_id` that is not yet in use for that sample:

```{r fig.small=TRUE, eval=TRUE}
url <- "https://i.redd.it/3pw5uah7xo041.jpg"
ve <- addImg(ve, 
    sample_id = "section1", image_id = "pomeranian",
    imageSource = url, scaleFactor = NA_real_, load = TRUE)

img <- imgRaster(ve, 
    sample_id = "section1", 
    image_id = "pomeranian")
plot(img)
```

The `rmvImg()` function is more flexible in the specification of the `sample/image_id` arguments. Specifically, 

- `TRUE` is equivalent to *all*, e.g. `sample_id = "<sample>", image_id = TRUE` will drop all images for a given sample.
- `NULL` defaults to the first entry available, e.g., `sample_id = "<sample>", image_id = NULL` will drop the first image for a given sample.

For example, `sample_id,image_id = TRUE,TRUE` will specify all images; `NULL,NULL` corresponds to the first image entry in the `imgData`; `TRUE,NULL` equals the first image for all samples; and `NULL,TRUE` matches all images for the first sample.

Here, we remove `section1`'s `pomeranian` image added in the previous code chunk; the image is now completely gone from the `imgData`:

```{r}
imgData(ve <- rmvImg(ve, "section1", "pomeranian"))
```

# `colData` replacement

<!-- While storing of `sample_id`s, the `in_tissue` indicator, and spatial `xy_coords` inside the `SpatialExperiment`'s `colData` enables directly accessibility via the `colData` and `$` accessors, these fields are protected against arbitrary modification. This affects replacement operations to the following effects: -->

<!-- Renaming is generally not permitted: -->

<!-- ```{r error = TRUE} -->
<!-- names(colData(ve))[1] <- "a" -->
<!-- ``` -->

Replacement of `sample_id`s is permitted provided that 

i. the number of unique sample identifiers is retained
ii. newly provided sample identifiers are a one-to-one mapping

```{r error = TRUE}
ve$sample_id <- sample(c("a", "b", "c"), ncol(ve), TRUE)
ve$sample_id <- sample(c("a", "b"), ncol(ve), TRUE)
```

Valid replacement will be propagated to the `imgData`:
  
```{r}
tmp <- ve
i <- as.numeric(factor(ve$sample_id))
tmp$sample_id <- c("sample1", "sample2")[i]
imgData(tmp)
```

The `in_tissue` field may be modified provided that the former is a logical vector

```{r error = TRUE}
ve$in_tissue <- "x"
```

`colData(.) <- NULL` will retain only the `sample_id`, `in_tissue` and `xy_coords` fields:

```{r}
names(colData(ve))
colData(ve) <- NULL
names(colData(ve))
```

# Visualization

For visualization examples (previous plots too) please see the [10xVisiumDemo](https://github.com/drighelli/SpatialAnalysisWorkflows/blob/master/10xVisiumDemo.Rmd)

# Session information {- .smaller}

```{r session-info}
sessionInfo()
```
